lib/state/preferences/reducer.ts

```diff
@@ -0,0 +1,30 @@
+import * as A from '../action-types';
+
+type Preferences = {
+  analytics_enabled: boolean | null;
+};
+
+const uninitializedPreferences = {
+  analytics_enabled: null,
+};
+
+const preferences: A.Reducer<Preferences> = (
+  state = uninitializedPreferences,
+  action
+) => {
+  switch (action.type) {
+    case 'TOGGLE_ANALYTICS':
+      return {
+        ...state,
+        analytics_enabled:
+          typeof action.sendAnalytics !== 'undefined' ||
+          !state.analytics_enabled,
+      };
+    case 'ANALYTICS_REMOTE_UPDATE':
+      return { ...state, analytics_enabled: action.sendAnalytics };
```

Comment:
two points here:

we don't have to accept this snake-case property in our app. we think about the properties as camelCase. we can preserve that by making a parsing step whenever the data comes in, and whenever the data goes out.

```ts
return { ...state, analyticsEnabled: action.sendAnalytics };
```

secondly, and explaining why some of this is confusing, it looks like we're trying to operate at the `analyticsEnabled` level _and_ at the `preferences` level at the same time. this is confusing.

_If_ we are going to treat this like any normal Simperium object we should swap it out here…

```ts
case 'PREFERENCES_REMOTE_UPDATE':
	return action.preferences;
```

…but otherwise we might want to just treat it as a simpler object, just the analytics…

```ts
const sendAnalytics: A.Reducer<boolean | null> = (state = null, action) => {
	switch (action.type) {
		case 'TOGGLE_ANALYTICS':
			return action.sendAnalytics ?? !state;

		case 'PREFERENCES_REMOTE_UPDATE':
			return action.data.sendAnalytics;
	}
}
```

We don't have to care about all the preferences object, but we should make sure we don't handicap ourselves from the onset.

```ts
const parsePreferences = ( { send_analytics, ...preferences } ) => ( {
	...preferences,
	sendAnalytics: send_analytics,
} );

const unparsePreferences = ( { sendAnalytics, ...preferences } ) => ( {
	...preferences,
	send_analytics: sendAnalytics
} );

// app.tsx
preferencesBucket.on('index', () => {
    preferencesBucket.get('preferences-key', (error, preferences) => {
        if (error || !preferences.data.hasOwnProperty('analytics_enabled')) {
            return;
        }

        this.props.remotePreferencesUpdate( parsePreferences( preferences.data ) );
    });
});

preferencesBucket.on('update', (entityId, data) => {
	if (
		'preferences-key' !== entityId ||
		!data.hasOwnProperty('analytics_enabled')
	) {
		return;
	}

	this.props.remotePreferencesUpdate( parsePreferences( data ) );
});

// middleware.ts
case 'TOGGLE_ANALYTICS':
	buckets.preferences.update( 'preferences-key', unparse( nextState.preferences ) );
```

Reply:
the naming thing isn't an issue with the `note` bucket because the names are already in the form we expect, otherwise we would do that there too. eventually we will probably want this parsing step anyway so we can stop doing things like `if ( note.data.systemTags.includes( 'pinned' ) )` and instead reliably rely on `if ( note.isPinned )`

these are enhancements not necessary yet, but in the case of preferences something felt more pertinent to me when I reviewed them